internal/envoy/versions_test.go

```diff
@@ -0,0 +1,91 @@
+// Copyright 2021 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/func-e/internal/version"
+)
+
+func TestFuncEVersions_FindLatestPatch(t *testing.T) {
+	type testCase struct {
+		name     string
+		input    version.Version
+		versions map[version.Version]version.Release
+		want     version.Version
+	}
+
+	tests := []testCase{
+		{
+			name:  "zero",
+			input: "1.20",
+			versions: map[version.Version]version.Release{
+				"1.20.0": {},
+			},
+			want: "1.20.0",
+		},
+		{
+			name:  "upgradable",
+			input: "1.18",
+			versions: map[version.Version]version.Release{
+				"1.18.3": {},
+				"1.18.4": {},
+			},
+			want: "1.18.4",
+		},
+		{
+			name:  "notfound",
+			input: "1.1",
+			versions: map[version.Version]version.Release{
+				"1.20.0": {},
+			},
+			want: "",
+		},
+	}
+
+	ctx := context.Background()
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			tester := newFuncEVersionsTester(tc.versions)
+			have, err := tester.feV.FindLatestPatch(ctx, tc.input)
+			if tc.want == "" {
+				require.Error(t, fmt.Errorf("couldn't find latest version for %s", tc.input))
+			} else {
+				require.NoError(t, err)
+				require.Equal(t, have, tc.want)
+			}
+		})
+	}
+}
+
+type funcEVersionsTester struct {
+	feV funcEVersions
+}
+
+func newFuncEVersionsTester(versions map[version.Version]version.Release) funcEVersionsTester {
+	return funcEVersionsTester{
+		feV: funcEVersions{
+			// Override Envoy versions getter for testing purpose only.
+			getFunc: func(context.Context) (version.ReleaseVersions, error) {
+				return version.ReleaseVersions{Versions: versions}, nil
+			},
+		},
+	}
+}
```

Comment:
I think maybe we don't need getFunc by just embedding  funcEVersions into funcEVersionsTester and override Get directly like this? 😄 

```suggestion
type funcEVersionsTester struct {
	funcEVersions // embed!
	versions map[version.Version]version.Release
}

// Override version.FuncEVersions
func (f *funcEVersions) Get(ctx context.Context) (version.ReleaseVersions, error) {
	return version.ReleaseVersions{Versions: versions}, nil
}

func newFuncEVersionsTester(versions map[version.Version]version.Release) version.FuncEVersions {
	return funcEVersionsTester{
		feV: funcEVersions{}
		versions: versions,
	}
}
```

Reply:
I think I have tried this before. Calling `Get` inside "parent's" `FindLatestPatch` is not overridden, if we call it from the tester object. But not sure: https://play.golang.org/p/aXL3DdUxSAg.